sdk/storage/Azure.Storage.Common/src/Shared/UriExtensions.cs

```diff
@@ -25,6 +25,9 @@ public static Uri AppendToPath(this Uri uri, string segment)
             var builder = new UriBuilder(uri);
             var path = builder.Path;
             var seperator = (path.Length == 0 || path[path.Length - 1] != '/') ? "/" : "";
+            // In URLs, the percent sign is used to encode special characters, so if the segment
+            // has a percent sign in their URL path, we have to encode it before adding it to the path
+            segment = segment.Replace(Constants.PercentSign, Constants.EncodedPercentSign);
```

Comment:
I was looking at recordings for "non ascii" tests. It seems that we send non-ascii characters encoded as `\uXXXX`. Should we encode `%` sign same way? I.e. send `\u0025` ?
https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#resource-names
http://kellykjones.tripod.com/webtools/ascii_utf8_table.html

Other question. Do we know where encoding of other non-ascii characters take place? Maybe we could push that logic down there (unless this % encoding of % sign is just a storage thing).

Reply:
I tried to replace the `%` character with a `\u0025` (instead of the %25) and looks like it just converts it back into a `%` instead of `\u0025` sign and doesn't encode the `%` like we want it.

The encoding of non-ascii characters take place in the UriBuilder and when making the Uri object. The Uri and UriBuilder class will take in a Uri and encode it for us. The Uri and UriBuilder class assumes that when you give it an escaped/encoded character, it should unencode it, cause it assumes we gave it an absolute or already escaped URL. That's why we see this behavior happen when we hand it `%3A` and not when it sees other characters like a `space` it encodes it correctly. `%` signs just are just a special thing in URLs.